### PR TITLE
[bugfix] fix errors on update operations

### PR DESCRIPTION
**Pull request by line-o**
### Description:

By setting the allowed-fragmentation attribute value to `-1` any modification will trigger the defragmentation of the dom.dbx (AKA BTree pages).

This setting triggers the NPE described in #5273 on the very first call to

```xquery
update (replace|insert) $nodeA (with|into) $nodeB
```

Unfortunately, it also triggers a totally different class of NPEs in XQueryTrigger and other subsystems of the database.

<details>
  <summary>Failures before d145f29857be1cf7be5ab15226f8239b5e5aac3d</summary>

```
[ERROR] Failures:
[ERROR]   TriggerConfigTest.removeDocument:137 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.removeDocument:137 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.removeDocument:137 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.storeDocument:114 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.storeDocument:114 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.storeDocument:114 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.updateTriggers:183 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.updateTriggers:183 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TriggerConfigTest.updateTriggers:183 Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.collectionCreate:522 expected:<1> but was:<0>
[ERROR]   XQueryTriggerTest.collectionMove:626 expected:<1> but was:<0>
[ERROR]   XQueryTriggerTest.documentBinaryDelete:475 expected:<1> but was:<0>
[ERROR]   XQueryTriggerTest.documentDelete:381 expected:<1> but was:<0>
[ERROR]   RESTServiceTest.xUpdate:642 Server returned response code 500 expected:<200> but was:<500>
[ERROR]   IndexIntegrationTest.insertElement Multiple Failures (2 failures)
	org.xmldb.api.base.XMLDBException: Failed to store collection configuration: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
	java.lang.IllegalStateException: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
[ERROR]   IndexIntegrationTest.removeAttribute Multiple Failures (2 failures)
	org.xmldb.api.base.XMLDBException: Failed to store collection configuration: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
	java.lang.IllegalStateException: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
[ERROR]   IndexIntegrationTest.updateAttribute:134->run:81->lambda$updateAttribute$3:173->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119
  Unexpected method call IndexWorker.setDocument(/db/test/pathNs2.xml - <test>):
    IndexWorker.flush(): expected: 2, actual: 2
[ERROR] Errors:
[ERROR]   XQueryTrigger2Test.collectionCopy:476 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.collectionCreate:448 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.collectionDelete:559 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.collectionMove:519 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.documentBinaryCreate:384 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.documentBinaryDelete:413 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.documentCreate:284 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.documentDelete:346 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTrigger2Test.documentUpdate:309 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.collectionCopy:550 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.collectionDelete:678 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.documentBinaryCreate:423 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.documentCreate:283 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTriggerTest.documentUpdate:317 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   AppendTest>AbstractUpdateTest.update:78->doUpdate:67 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   RangeIndexUpdateTest.updates:118 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   RemoveTest>AbstractUpdateTest.update:78->doUpdate:77 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   RenameTest>AbstractUpdateTest.update:78->doUpdate:72 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ReplaceTest>AbstractUpdateTest.update:78->doUpdate:69 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateAttributeTest>AbstractUpdateTest.update:78->doUpdate:70 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateRecoverTest.storeAndRead:88->store:153 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateRecoverTest.storeAndReadXmldb:100->xmldbStore:336 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateTest>AbstractUpdateTest.update:78->doUpdate:71 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   SerializationTest.xqueryUpdateNsTest:154 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   SerializationTest.xqueryUpdateNsTest:154 » XMLDB Failed to invoke method queryPT in class org.exist.xmlrpc.RpcConnection: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ComplexUpdateTest>ConcurrentTestBase.concurrent:139 » Execution java.lang.NullPointerException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ConcurrentAttrUpdateTest>ConcurrentTestBase.concurrent:139 » Execution java.lang.NullPointerException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ConcurrentQueryUpdateTest>ConcurrentTestBase.concurrent:139 » Execution org.xmldb.api.base.XMLDBException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ConcurrentXUpdateTest>ConcurrentTestBase.concurrent:139 » Execution java.lang.NullPointerException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   TextUpdateTest>ConcurrentTestBase.concurrent:139 » Execution java.lang.NullPointerException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ValueIndexUpdateTest>ConcurrentTestBase.concurrent:139 » Execution java.lang.NullPointerException: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   DocumentUpdateTest.update:110->execQuery:156 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   DocumentUpdateTest.updateAttribute:149->execQuery:156 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   NamespaceUpdateTest.updateAttribute:64 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   QueryPoolTest.differentQueries:54 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ValueIndexByQNameTest>ValueIndexTest.updates:360 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ValueIndexByQNameTest>ValueIndexTest.updatesQName:392 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ValueIndexTest.updates:360 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   ValueIndexTest.updatesQName:392 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XPathQueryTest.ids_persistent:1617->queryResource:2279->queryResource:2296 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XPathQueryTest.ids_persistent:1617->queryResource:2279->queryResource:2296 » XMLDB Failed to invoke method queryPT in class org.exist.xmlrpc.RpcConnection: Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryTest.attribute_1691177:1788 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.appendAttributes:105->append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.appendCDATA:432 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.attrUpdate:413 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.insertAfter:221 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.insertAttrib:462 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.insertBefore:167 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.remove:315->append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.rename:335->append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.replace:364->append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQueryUpdateTest.update:263->append:81 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateInsertTest.insertInMemoryDocument:102->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateInsertTest.insertNamespacedAttribute:46->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateInsertTest.insertPrecedingAttribute:83->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateReplaceTest.replaceFirstChildWhereParentHasAttribute:157 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateReplaceTest.replaceFirstChildWhereParentHasNoAttributes:89 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateReplaceTest.replaceOnlyChildWhereParentHasAttribute:123 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateReplaceTest.replaceOnlyChildWhereParentHasNoAttributes:55 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateValueTest.updateAttributeInNamespacedElement:54->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   UpdateValueTest.updateNamespacedAttribute:42->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119 » XMLDB Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   RemoveAppendTest.appendRemove:91->append:116 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   StressTest.stressTest:73->insertTags:98 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XUpdateTest.xupdate:132->updateDocument:185 » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   CoreTests » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
[ERROR]   XQuery3Tests » NullPointer Cannot invoke "org.exist.storage.BrokerPool.getSecurityManager()" because "pool" is null
```
</details>
<details>
  <summary>Failures after adding the fix for #5273</summary>

```
[ERROR]   IndexIntegrationTest.insertElement Multiple Failures (2 failures)
	org.xmldb.api.base.XMLDBException: Failed to store collection configuration: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
	java.lang.IllegalStateException: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
[ERROR]   IndexIntegrationTest.removeAttribute Multiple Failures (2 failures)
	org.xmldb.api.base.XMLDBException: Failed to store collection configuration: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
	java.lang.IllegalStateException: missing behavior definition for the preceding method call:
IndexWorker.getListener()
Usage is: expect(a.foo()).andXXX()
[ERROR]   IndexIntegrationTest.updateAttribute:134->run:81->lambda$updateAttribute$3:173->AbstractTestUpdate.queryResource:105->AbstractTestUpdate.queryResource:119
  Unexpected method call IndexWorker.setDocument(/db/test/pathNs2.xml - <test>):
    IndexWorker.flush(): expected: 2, actual: 2
[ERROR] Errors:
[ERROR]   XQueryTriggerTest.collectionCopy:550 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.collectionCreate:513 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.collectionDelete:675 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.collectionMove:614 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.documentBinaryCreate:423 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.documentBinaryDelete:463 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.documentCreate:283 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.documentDelete:369 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.documentUpdate:317 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
[ERROR]   XQueryTriggerTest.storeDocumentInvalidTriggerForPrepare:719 » NullPointer Cannot invoke "org.exist.storage.btree.Value.data()" because "oldVal" is null
```
</details>

-- UPDATE --

The first attempt to fix modifying files did surface a new issue. A non-DBA user triggering the defragmentation of an XML resource would lead to a PermissionDeniedException. The patch was updated to copy the documents permissions instead of applying the DB's default permissions (see 1ca99018758d6ae32fa68f725a1643119927630a).

### CODE REVIEW

**Comment by adamretter** [comment]
> By setting the allowed-fragmentation attribute value to -1 any modification will trigger the defragmentation of the structural(?) index (AKA BTree).

B-Tree's don't become fragmented!

File: exist-core/src/test/resources-filtered/conf.xml
Comment (adamretter) [inline exist-core/src/test/resources-filtered/conf.xml:932]: You don't need to modify the config for every test, and neither should you! You can simply override this in your JUnit test that checks for this by passing configuration properties to the constructor of your EmbeddedExistServer or EmbeddedExistXmldbServer.

**Comment by wolfgangmm** [comment]
> > By setting the allowed-fragmentation attribute value to -1 any modification will trigger the defragmentation of the structural(?) index (AKA BTree).
> 
> B-Tree's don't become fragmented!

... though the data pages in `dom.dbx` do.

File: exist-core/src/test/resources-filtered/conf.xml
Comment (line-o) [reply exist-core/src/test/resources-filtered/conf.xml:932]: I do feel it would be good to run our tests in a way that will ensure they work even when they trigger a defragmentation run.

**Comment by wolfgangmm** [comment]
> @wolfgangmm is it the dom.dbx that gets rewritten on `defragXMLresource`? I would clarify that in the description then.

Inserting something in the middle of a document may cause a page split in the data section of `dom.dbx` (where the dom nodes are stored sequentially). `defragXMLResource` was introduced to compact the document again when it grows too much. Indexes tracking the logical storage address of particular nodes also have to be updated afterwards.

File: exist-core/src/test/resources-filtered/conf.xml
Comment (adamretter) [reply exist-core/src/test/resources-filtered/conf.xml:932]: Changing the behaviour of every test for this single purpose is not appropriate.

**Comment by adamretter** [comment]
> ... though the data pages in dom.dbx do.

@wolfgangmm If I recall aren't there also other `.dbx` files that can experience page splits too?

> Inserting something in the middle of a document may cause a page split

Isn't it actually dependent on the page size rather than the document size. i.e. inserting anywhere in a document where it exceeds a page boundary?

File: exist-core/src/test/resources-filtered/conf.xml
Comment (line-o) [reply exist-core/src/test/resources-filtered/conf.xml:932]: It serves a single purpose: highlight all areas where the defragmentation of split documents will cause problems.
And as you can see from the list of failing tests XQueryTrigger might still be affected and this needs further investigation.

Comment (adamretter) [reply exist-core/src/test/resources-filtered/conf.xml:932]: > It serves a single purpose

Quite the opposite. It mixes concerns... and it will trigger unnecessary defragmentation in all other tests that cause a page slit. This means side-effects and slow-down in the test suite. It isn't needed. Please just add a test for your issue.

Comment (line-o) [reply exist-core/src/test/resources-filtered/conf.xml:932]: The thing is, creating a single test-case by creating an instance of EmbeddedExistXmldbServer with the changed setting - as you proposed - was my first attempt. The Properties one uses to set fragmentation limit must be of type string and cannot be an integer value. They are also not casted to an integer where needed. 

Conclusion: It does not work that way.

Comment (adamretter) [reply exist-core/src/test/resources-filtered/conf.xml:932]: > The Properties one uses to set fragmentation limit must be of type string and cannot be an integer value

@line-o That's not correct, you can set any type of value.

> Conclusion: It does not work that way.

I am not sure what you tried as I can't see your code. I created a quick skeleton for you that works just fine:

```java
package org.exist.storage.btree;

import org.exist.storage.BrokerPool;
import org.exist.test.ExistEmbeddedServer;
import org.exist.util.Configuration;
import org.junit.Before;
import org.junit.ClassRule;
import org.junit.Test;

import java.util.Properties;

import static org.exist.storage.DBBroker.PROPERTY_XUPDATE_FRAGMENTATION_FACTOR;

public class PagedDefragmentationTest {

  private static final Properties configProperties = new Properties();
  static {
    configProperties.put(PROPERTY_XUPDATE_FRAGMENTATION_FACTOR, Integer.valueOf(-1));
  }

  @ClassRule
  public static final ExistEmbeddedServer existEmbeddedServer = new ExistEmbeddedServer(configProperties, true, true);

  // NOTE(AR) just temporary to show that config of Integer properties works within a test context
  @Before
  public void printStatus() {
    final BrokerPool brokerPool = existEmbeddedServer.getBrokerPool();
    System.out.println("Operational=" + brokerPool.isOperational());
    System.out.println("Configured=" + brokerPool.isInstanceConfigured());
    final Configuration configuration = brokerPool.getConfiguration();
    final Object configuredValue = configuration.getProperty(PROPERTY_XUPDATE_FRAGMENTATION_FACTOR);
    System.out.println("Configuration: " + PROPERTY_XUPDATE_FRAGMENTATION_FACTOR + "=" + configuredValue + " (type: " + configuredValue.getClass().getSimpleName() + ")");
  }

  @Test
  public void yourTest() {
    // TODO
  }

}

```

**Comment by dizzzz** [comment]
we had a good session yestersay (Patrick, Me, Me), exciting to see the PR extended soonish...

File: exist-core/src/test/resources-filtered/conf.xml
Comment (line-o) [reply exist-core/src/test/resources-filtered/conf.xml]: Thanks, I was missing the put method on the properties. The test is now in place.
Regardless, I still am convinced that XQueryTriggerTests do indicate there is a problem when defragmentation is triggered.

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (line-o) [reply exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:42]: Well, it's in line with all the other tests next to it. I would certainly change them all to follow the naming conventions.
I would rather do that in one step to all tests that are not following them all at once.

**Review by reinhapa — APPROVED** [review]
LGTM

**Review by dizzzz — APPROVED** [review]

**Comment by dizzzz** [comment]
LGTM
nice job, good find, happy to see the issue fixed.!

**Comment by dizzzz** [comment]
Port to 6.x is needed

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:47]: 1. These should not be class members, also they don't need to be. 2. The `testCollection` has not been closed when it is finished with, so this test may leak resources.

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:59]: 1. The XML Resource has not been closed after it is finished with, so this test may leak resources.

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:73]: This method is not needed at all because `useTemporaryStorage = true` is set at the start of this class.

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (adamretter) [inline exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:80]: Does the `ResourceSet` also need to be closed?

**Comment by adamretter** [comment]
It seems the changes to fix the initial review went in less than 24 hours ago, and a second review was requested less than 5 hours ago. As someone who raised issues with the initial PR, and provided code to help get this PR working, it would have been nice to have had a little more time to have the opportunity do a second review before it was merged!

Ignoring the still unanswered questions/observations about when page-splits occur.
I did a quick look over the merged code, I think unfortunately the new code also has issues whereby the test that was added may leak resources. IMHO this should not have been merged yet, however moving forward, I think it would be good for the author to send a follow-up PR to address these leaks.

**Comment by line-o** [comment]
@ada Please open a separate issue or PR on the changes you would like to see in develop.

File: exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java
Comment (line-o) [reply exist-core/src/test/java/org/exist/xquery/update/UpdateInsertTriggersDefrag.java:73]: This was copied from `AbstractTestUpdate` is it wrong there too?

**Comment by adamretter** [comment]
I have sent a PR to address the issues introduced by this merged PR here - https://github.com/eXist-db/exist/pull/5503

**Pull request merged**
